base: Call rpmReadConfigFiles() earlier

m-blaha opened the pull request:
Currently, libdnf5 initializes librpm only when detecting `arch` or
`releasever` variables. In case both these values are provided using
command line options (--releasever, --forcearch), `rpmReadConfigFiles()`
is not called at all, resulting in uninitialized librpm.

This patch moves `init_lib_rpm()` call to the `Base` class constructor,
ensuring that `rpmReadConfigFiles()` is always called.

Resolves: https://github.com/rpm-software-management/dnf5/issues/855

j-mracek commented:
LGTM. @m-blaha Do you think whether is it possible to create a test for out CI?

m-blaha commented:
I can create a test for this particular occurrence of working with initialized rpm - make sure that `dnf5 install --releasever=xx --forcearch=yy` works correctly.

m-blaha commented:
PR with test: https://github.com/rpm-software-management/ci-dnf-stack/pull/1376

pmatilai commented on libdnf5/base/base.cpp:
```diff
@@ -82,13 +84,25 @@ namespace libdnf5 {
+static void init_lib_rpm() {
+    static bool lib_rpm_initiated{false};
+    if (!lib_rpm_initiated) {
+        if (rpmReadConfigFiles(nullptr, nullptr) != 0) {
```
I don't know what exactly --forcearch does inside libdnf, but since it's obviously about overriding the detected arch: it might make sense to pass it down to rpm too, ie as the second argument of rpmReadConfigFiles().

m-blaha replied:
Hm, that might make things more complicated than I assumed :)
Is it possible to call `rpmReadConfigFiles()` multiple times? Currently the function is called very early, even before the `--forcearch` switch is processed (for example because detecting the releasever needs to query rpm database). So if it was possible, we could use the second call of `rpmReadConfigFiles` to "adjust" the architecture during processing of `--forcearch` switch. If this is not possible (which I assume), we need to handle arch and releasever variables differently and initialize rpmdb later.
Then we might run into kind of chicken and egg problem - reading the configuration needs variables (thus to have rpm initialized) and initialization of rpm needs `arch` value which is obtained from configuration (or --forcearch parameter).

pmatilai replied:
Calling rpmReadConfigFiles() multiple times is perfectly okay (rpmbuild itself does this), just remember to call rpmFreeRpmrc() for each round.

m-blaha commented:
@j-mracek can you please review again?

- librpm is now initialized a bit later - in base.setup(). We need to do it after commandline argument parsing (to be able to extract --forcearch value), but before loading variables.
- added passing --forcearch value also to `rpmReadConfigFiles()` (according to Panu's comment)
- added handling of unknown arch used in --forcearch with explanatory message to the user

j-mracek commented on include/libdnf5/rpm/arch.hpp:
```diff
@@ -0,0 +1,37 @@
+
+#include "libdnf5/common/exception.hpp"
+
+#include <sstream>
```
May be I overlooked something, but it looks like that 
```
#include "libdnf5/common/exception.hpp"

#include <sstream>
```
are not required.

m-blaha replied:
Oh, this copy and paste (mainly because of this ugly license text). Looking forward to have short, neat SPDX identifiers instead...

m-blaha replied:
Fixed.

j-mracek commented on include/libdnf5/rpm/arch.hpp:
```diff
@@ -0,0 +1,34 @@
+
+namespace libdnf5::rpm {
+
+std::vector<std::string> get_supported_arches();
```
May be we need it, but I am confused another thing - structure from which the vector is generated.

It is c structure and it is not nice for our usage. The construct is <basearch>, <arches> therefore you have to iterate over the whole structure. If we would like to improve the search then we can create std::unordered_map<string, string> where key will be <arch> and the value <basearch>. The problem of this approach is that map is bigger then the original structure, and <Basearch> strings will be present multiple times, but it can be improved by another approach
```
std::unordered_map<string, int>  // <arch>, <index>
std::array<std::string>, where index could be used to retrieve the Basearch.
```
@m-blaha What do you think? @jrohel May I ask you for your opinion?

m-blaha replied:
It's used by dnf5 to generate an error message if unsupported arch is used.

m-blaha replied:
Regarding the `ARCH_MAP` data structure, you are right. I guess it was copied from libdnf where probably the C structure was needed. In dnf5 we can replace it by C++ structure (like std::unordered_map).

j-mracek replied:
may be `std::map` could be better if you want to have get_supported_arches ordered without ordering.

jrohel replied:
- I would prefer `std::map` over `std::unordered_map` for these tiny maps.
- And inside `std::string_view`. No need to copy those strings to the heap.

But it will still take up more memory than the original code.

m-blaha replied:
@j-mracek do you want me to change also the underlying `ARCH_MAP` structure, or is TODO sufficient for this PR?

jrohel replied:
@m-blaha Why do you actually want to replace the C structure? How many times do we call `get_base_arch`? Is it worth it to construct a map?
The cost to create a map will be several times higher than calling `get_base_arch`.

Is the requirement or do we need to intensively call `get_base_arch`? (Note: If so, then code can be written to create a sorted reverse mapping (array of native -> base structs) with little cost directly at compile time.)

j-mracek replied:
I would like to clarify internal stricture, because there is a method (public) which depends on. The method returns a vector with strings.

j-mracek replied:
I am fine even with TODO.

j-mracek replied:
@jrohel - I was thinking that the map is created during compilation. Or did I overlooked something?

j-mracek commented on libdnf5/base/base.cpp:
```diff
@@ -221,7 +233,13 @@ void Base::setup() {
+    if (vars->contains("arch")) {
+        arch = vars->get_value("arch").c_str();
+    }
+    init_lib_rpm(arch);
```
May I ask you to add a note that when locking of vars will be available we should lock all vars or arch variable.

Are you sure that the order of following lines is correct? I somehow have feeling that we should first load and then init rpm with architecture.
```
init_lib_rpm(arch);
vars->load(vars_installroot, config.get_varsdir_option().get_value());
```

m-blaha replied:
The thing is, that during `vars->(load)` the architecture and releasever are also detected (if missing). And this detection needs rpm correctly initialized.

m-blaha replied:
I've added a comment on it.

j-mracek commented on libdnf5/base/base.cpp:
```diff
@@ -183,7 +195,17 @@ void Base::setup() {
+    if (vars->contains("arch")) {
+        arch = vars->get_value("arch").c_str();
+    }
+    init_lib_rpm(arch);
```
I have additional  comment. If I am not mistaken, we pass rpm an architecture, but in this time we can still use a different architecture, because architecture might be modify or set by `vars->load(vars_installroot, config.get_varsdir_option().get_value());`. It means that we might use for POOL a different architecture that we provide to RPM. @pmatilai If we will use the same architecture in libsolv pool and we will initialize rpm config `rpmReadConfigFiles(nullptr, arch)` with the same architecture, could we stop use `RPMPROB_FILTER_IGNOREARCH` for transaction where architecture used for transaction calculation differs from the system architecture.

m-blaha replied:
Gosh, that's right! Thanks for catching it. So I'll move the code back to vars.cpp and just make sure that the code is called unconditionally.

m-blaha replied:
@j-mracek the issue with architecture provided by vars file should be resolved. Now the rpm is initialized during Vars::detect_vars() call.

j-mracek commented on include/libdnf5/rpm/arch.hpp:
```diff
@@ -0,0 +1,34 @@
+
+namespace libdnf5::rpm {
+
+std::vector<std::string> get_supported_arches();
```
There is somehow related issue - https://github.com/rpm-software-management/dnf5/issues/600

j-mracek commented on libdnf5/conf/vars.cpp:
```diff
@@ -433,12 +377,21 @@ void Vars::load(const std::string & installroot, const std::vector<std::string>
+        arch = get_value("arch").c_str();
+    }
+    init_lib_rpm(arch);
+
```
This is very minor comment - you already have an information for test whether arch is set or not therefore it is not required to ask set_lazy to check it again. 
```
    set_lazy(
        "arch", []() -> auto { return std::make_unique<std::string>(detect_arch()); }, Priority::AUTO);
```

j-mracek commented on libdnf5/conf/vars.cpp:
```diff
@@ -433,12 +377,21 @@ void Vars::load(const std::string & installroot, const std::vector<std::string>
+    if (contains("arch")) {
+        arch = get_value("arch").c_str();
+    }
+    init_lib_rpm(arch);
```
I think this is the perfect place. Thank you.

j-mracek commented:
LGTM
